[WIP] OSDOCS-3417: separate out sharding docs (#49136)
openshift/openshift-docs: modules/nw-ingress-sharding.adoc

// Module included in the following assemblies:
//
// * ingress-operator.adoc
// * networking/ingress-sharding.adoc

:_content-type: CONCEPT
[id="nw-ingress-sharding_{context}"]
= Ingress Controller sharding

> Ingress sharding, also known as router sharding, lets you distribute a set of routes across multiple routers by adding labels to routes, namespaces, or both. The Ingress Controller uses a corresponding set of selectors to admit only the routes with a specified label. Each Ingress shard comprises the routes that are filitered using a given selection expression.

Reviewer: Remember to check for spelling. "filitered"

As the primary mechanism for traffic to enter the cluster, the demands on the Ingress Controller can be significant. As a cluster administrator, you can shard the routes to:

* Balance Ingress Controllers, or routers, with several routes to speed up responses to changes.
* Allocate certain routes to have different reliability guarantees than other routes.
* Allow certain Ingress Controllers to have different policies defined.
* Allow only specific routes to use additional features.
* Expose different routes on different addresses so that internal and external users can see different routes, for example.
* Transfer traffic from one version of an application to another during a blue green deployment.

When Ingress Controllers are sharded, a given route is admitted to zero or more Ingress Controllers in the group. A route's status describes whether an Ingress Controller has admitted it or not. An Ingress Controller will only admit a route if it is unique to its shard.

An Ingress Controller can use three sharding methods:

> . Adding only a namespace selector to the Ingress Controller, so that all routes in a namespace with labels that match the namespace selector are in the Ingress shard.

Reviewer: ISG: Is the sequence significant? If not, use an unordered list.

. Adding only a route selector to the Ingress Controller, so that all routes with labels that match the route selector are in the Ingress shard.

. Adding both a namespace selector and route selector to the Ingress Controller, so that routes with labels that match the route selector in a namespace with labels that match the namespace selector are in the Ingress shard.

> This lets you distribute subsets of routes over multiple Ingress Controllers. These subsets can be non-overlapping, also called _traditional_ sharding, or overlapping, otherwise known as _overlapped_ sharding.

Reviewer: This what? Read ambiguously to me.

> == Traditional sharding example

Reviewer: These examples could be broken out into reference modules if you liked.

Author: This is a good idea, thanks! I think they could be tweaked a little more, so I'll open a followup PR for modularization

[source,yaml]
----
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: finops-router
    namespace: openshift-ingress-operator
  spec:
    namespaceSelector:
      matchLabels:
        name:
          - finance
          - ops
----

A second Ingress Controller `dev-router` is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `dev`:

.Example YAML definition for `dev-router`
[source,yaml]
----
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: dev-router
    namespace: openshift-ingress-operator
  spec:
    namespaceSelector:
      matchLabels:
        name: dev
----

If all application routes are in separate namespaces, each labeled with `name:finance`, `name:ops`, and `name:dev` respectively, this configuration effectively distributes your routes between the two Ingress Controllers. {product-title} routes for console, authentication, and other purposes should not be handled.

In the above scenario, sharding becomes a special case of partitioning, with no overlapping subsets. Routes are divided between router shards.

[WARNING]
====

> The `default` Ingress Controller will continue to serve all routes unless the `namespaceSelector` or `routeSelector` fields contain routes that are meant for exclusion. See this link:https://access.redhat.com/solutions/5097511[Red Hat Knowledgebase solution] and the section "Sharding the default Ingress Controller" for more information on how to exclude routes from the default Ingress Controller.

Reviewer: Does this work just as well?: "will continue to serve" -> "serves"

Author: maybe, but it's a good point, I'll update it. I'm trying to convey that even if you create custom Ingress Controllers to handle sharding, it doesn't change the behavior of the default Ingress Controller...the custom ones use kind of an inclusive logic in their config and the default uses more of an exclusive logic. So maybe "continues to serve" or something

Reviewer: Could just be explicit about it in the admonition, unless that information is definitely going to be in the reader's head at this point. 

It's admittedly different doing a peer review in a code editor than it is following docs as a real user, so I could be off base. :+1:

====

== Overlapped sharding example

> In addition to `finops-router` and `dev-router` in the example above, you also have `devops-router`, that is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `dev` and `ops`:

Reviewer: Non-restrictive clause here, so "which is configured"

.Example YAML definition for `devops-router`
[source,yaml]
----
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: devops-router
    namespace: openshift-ingress-operator
  spec:
    namespaceSelector:
      matchLabels:
        name:
          - dev
          - ops
----

> The routes in the namespaces labeled `name:dev` and `name:ops` are now serviced by two different Ingress Controllers. This becomes a case in which you have defined overlapping subsets of routes.

Reviewer: "This becomes a case in which you have defined" -> "With this configuration, you have" 

Assuming I have the idea right.

* Allow certain Ingress Controllers to have different policies defined.
* Allow only specific routes to use additional features.
* Expose different routes on different addresses so that internal and external users can see different routes, for example.
* Transfer traffic from one version of an application to another during a blue green deployment.

> When Ingress Controllers are sharded, a given route is bound to zero or more routers in the group. The route binding ensures uniqueness of the route across the shard. Uniqueness allows secure and non-secure versions of the same route to exist within a single shard. This implies that routes now have a visible life cycle that moves from created to bound to active.

Author: This text is mostly lifted from the 3.10 docs, but if there's a better way to describe things we should definitely do that. @gcs278 any ideas?

  spec:
    namespaceSelector:
      matchLabels:
        name: dev
----

If all routes are in the namespaces labeled `name:finance`, `name:ops`, and `name:dev`, then this configuration effectively distributes your routes between the two router deployments.

In the above scenario, sharding becomes a special case of partitioning, with no overlapping subsets. Routes are divided between router shards.

== Overlapped sharding example

In addition to `finops-router` and `dev-router` in the example above, you also have `devops-router`, that is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `dev` and `ops`:

.Example YAML definition for `devops-router`
[source,yaml]
----
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: devops-router
    namespace: openshift-ingress-operator
  spec:
    namespaceSelector:
      matchLabels:
        name:
          - dev
          - ops
----
The routes in the namespaces labeled `name:dev` and `name:ops` are now serviced by two different router deployments. This becomes a case in which you have defined overlapping subsets of routes.

> Additionally, having overlapping subsets of routes lets you creat more complex routing rules, allowing the diversion of higher priority traffic to the dedicated `finops-router` while sending lower priority traffic to `devops-router`.

Reviewer 2: Small typo
```suggestion
Additionally, having overlapping subsets of routes lets you create more complex routing rules, allowing the diversion of higher priority traffic to the dedicated `finops-router` while sending lower priority traffic to `devops-router`.
```

> Additionally, having overlapping subsets of routes lets you creat more complex routing rules, allowing the diversion of higher priority traffic to the dedicated `finops-router` while sending lower priority traffic to `devops-router`.

Author: @candita it's in the second "Example" subsection here: https://docs.openshift.com/container-platform/3.10/install_config/router/default_haproxy_router.html#using-router-shards

* Allow certain Ingress Controllers to have different policies defined.
* Allow only specific routes to use additional features.
* Expose different routes on different addresses so that internal and external users can see different routes, for example.
* Transfer traffic from one version of an application to another during a blue green deployment.

When Ingress Controllers are sharded, a given route is bound to zero or more routers in the group. The route binding ensures uniqueness of the route across the shard. Uniqueness allows secure and non-secure versions of the same route to exist within a single shard. This implies that routes now have a visible life cycle that moves from created to bound to active.

> 

Reviewer 3: Also, related to the comment on uniqueness, we noticed that the idea of route "uniqueness", though not specific to sharding, it not really explained anywhere in the documentation. The idea of route uniqueness is that a route is unique if has one of characteristics different:
1. TLS Termination
2. Hostname
3. Path

I don't think this concept is address in documentation, but I think that's probably something we can address later and not in this PR.

Author: @gcs278 thanks for this, do you have more information on the topic? I can create a JIRA ticket to address this as a continuous improvement task (CIT) now that we're post GA 🙂

Reviewer 3: Sounds good, I think that's about it, but if you make the Jira Ticket, I can help fill in any holes (or I can get Miciah to if I can't).

Ingress Controller can use three sharding methods:

> . Adding only a namespace selector to the Ingress Controller, so that all routes in that namespace are in the Ingress shard.

Reviewer 3: Wording seems to suggest the fact labels aren't involved.

```suggestion
. Adding only a namespace selector to the Ingress Controller, so that all routes in a namespace with labels that match the namespace selector are in the Ingress shard.
```

> . Adding only a route selector to the Ingress Controller, so that all routes matching that route selector are in the Ingress shard.

Reviewer 3: ```suggestion
. Adding only a route selector to the Ingress Controller, so that all routes with labels that match the route selector are in the Ingress shard.
```

> . Adding both a namespace selector and route selector to the Ingress Controller, so that routes in both the given namespace and which match the route selector are in the Ingress shard.

Reviewer 3: Aiming for wording consistency.

```suggestion
. Adding both a namespace selector and route selector to the Ingress Controller, so that routes with labels that match the route selector in a namespace with labels that match the namespace selector are in the Ingress shard.
```

> This lets you distribute subsets of routes over multiple router deployments. These subsets can be non-overlapping, also called _traditional_ sharding, or overlapping, otherwise known as _overlapped_ sharding.

Reviewer 3: ```suggestion
This lets you distribute subsets of routes over multiple Ingress Controllers. These subsets can be non-overlapping, also called _traditional_ sharding, or overlapping, otherwise known as _overlapped_ sharding.
```

== Traditional sharding example

> A router deployment `finops-router` is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `finance` and `ops`:

Reviewer 3: ```suggestion
An Ingress Controller `finops-router` is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `finance` and `ops`:
```

.Example YAML definition for `finops-router`
[source,yaml]
----
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: finops-router
    namespace: openshift-ingress-operator
  spec:
    namespaceSelector:
      matchLabels:
        name:
          - finance
          - ops
----

> A second router deployment `dev-router` is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `dev`:

Reviewer 3: ```suggestion
A second Ingress Controller `dev-router` is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `dev`:
```

.Example YAML definition for `dev-router`
[source,yaml]
----
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: dev-router
    namespace: openshift-ingress-operator
  spec:
    namespaceSelector:
      matchLabels:
        name: dev
----

If all application routes are in the namespaces labeled `name:finance`, `name:ops`, and `name:dev`, then this configuration effectively distributes your routes between the two router deployments. {product-title} routes for console, authentication, and other purposes should not be handled.

In the above scenario, sharding becomes a special case of partitioning, with no overlapping subsets. Routes are divided between router shards.

[WARNING]
====

> The `default` router deployment will continue to serve all routes unless the `namespaceSelector` or `routeSelector` fields contain routes that are meant for exclusion. See this link:https://access.redhat.com/solutions/5097511[knowledgebase solution] for more information on how to exclude routes from the default Ingress Controller.

Reviewer 3: ```suggestion
The `default` Ingress Controller will continue to serve all routes unless the `namespaceSelector` or `routeSelector` fields contain routes that are meant for exclusion. See this link:https://access.redhat.com/solutions/5097511[knowledgebase solution] for more information on how to exclude routes from the default Ingress Controller.
```

====

== Overlapped sharding example

In addition to `finops-router` and `dev-router` in the example above, you also have `devops-router`, that is configured with the label selector `spec.namespaceSelector.matchLabels.name` set to `dev` and `ops`:

.Example YAML definition for `devops-router`
[source,yaml]
----
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: devops-router
    namespace: openshift-ingress-operator
  spec:
    namespaceSelector:
      matchLabels:
        name:
          - dev
          - ops
----

> The routes in the namespaces labeled `name:dev` and `name:ops` are now serviced by two different router deployments. This becomes a case in which you have defined overlapping subsets of routes.

Reviewer 3: ```suggestion
The routes in the namespaces labeled `name:dev` and `name:ops` are now serviced by two different Ingress Controller. This becomes a case in which you have defined overlapping subsets of routes.
```
